### PR TITLE
Bugfix: Several lisk-desktop identifiers are not unique

### LINKED ISSUE #5143: Several lisk-desktop identifiers are not unique
### Description

The lisk-desktop application stores names for accounts, networks, and bookmarks. All of these are not guaranteed to be unique.

image

### Recommendation

Guarantee that each of the account, network, and bookmark identifiers have to be unique. This will reduce the likelihood of user mistakes.

### Which version(s) does this affect? (Environment, OS, etc...)
3.0

### PR DESCRIPTION
**Pull request by ikem-legend**
### What was the problem?

This PR resolves #5143

### How was it solved?

Implemented unique identifiers for:
- [x] Accounts
- [x] Networks
- [x] Bookmarks

### How was it tested?

- Manually
Create a duplicate account name, bookmark or network. For example, create a custom network called `mainnet` or `devnet` and you should see an error stating it's a duplicate
- Unit tests

### CODE REVIEW

File: setup/jest.config.js [hunk 1]
```diff
@@ -132,7 +132,6 @@ module.exports = {
     'src/modules/common/components/timestamp/index.js',
```
Comment (oskarleonard) [inline setup/jest.config.js:132]: NRTTL
Account and network takes spaces and casing under consideration (which is the best approach regarding this i think).  While bookmarks dont, lets add that behavior for bookmarks as well.

File: setup/jest.config.js [hunk 2]
```diff
@@ -132,7 +132,6 @@ module.exports = {
...
     'src/utils/localJSONStorage.js',
```
Comment (oskarleonard) [inline setup/jest.config.js:133]: NRTTL
I can still get duplicate name using edit function for bookmarks and network (edit account is fine though)

File: src/modules/bookmark/utils.js [hunk 3]
```diff
@@ -26,17 +25,25 @@ export const validateBookmarks = (data) => {
...
  *  Checks the label and returns feedback
  *
+ * @param {String} token - LSK or any other token
  * @param {String} value - The label string to check
+ * @param {Object} bookmarks - List of bookmarks from Redux store
  * @param {Function} t - i18n function
  * @returns {String} - Feedback string. Empty string if the label is valid
  */
-export const validateBookmarkLabel = (value = '', t) => {
+export const validateBookmarkLabel = (token, value = '', bookmarks, t) => {
```
Comment (oskarleonard) [inline src/modules/bookmark/utils.js:41]: Why is token part of bookmark validation?

File: src/modules/network/components/DialogAddNetwork/DialogAddNetwork.js [hunk 4]
```diff
@@ -44,18 +45,19 @@ const DialogAddNetwork = () => {
...
-      (network) => network.name === values.name
-    );
-    const existingCustomNetworkServiceUrl = customNetworks.some(
-      (network) => network.serviceUrl === values.serviceUrl
+    const fullNetworkList = [...Object.values(networks), ...customNetworks];
+    const existingCustomNetworkName = fullNetworkList.some(
+      (network) => network.name.toLowerCase() === values.name.toLowerCase()
     );
+    const existingCustomNetworkServiceUrl = fullNetworkList
+      .filter((network) => network.serviceUrl !== 'http://localhost:9901')
```
Comment (oskarleonard) [inline src/modules/network/components/DialogAddNetwork/DialogAddNetwork.js]: Why this hard coding of 'http://localhost:9901' ?

File: src/modules/bookmark/utils.js (hunk 3 above)
Comment (ikem-legend) [reply src/modules/bookmark/utils.js:41]: Bookmarks are stored by token so it makes sense to pass in the active token

File: src/modules/network/components/DialogAddNetwork/DialogAddNetwork.js (hunk 4 above)
Comment (ikem-legend) [reply src/modules/network/components/DialogAddNetwork/DialogAddNetwork.js]: The localhost address is among the default list of addresses. If it's not filtered, then the user will be unable to add that address as a custom node

File: src/modules/account/components/EditAccount/EditAccountForm.js [hunk 5]
```diff
@@ -37,11 +37,20 @@ const EditAccountForm = ({ nextStep }) => {
...
   const formValues = watch();
   const [currentAccount] = useCurrentAccount();
+  const { accounts } = useAccounts();
 
   const onSubmit = async ({ accountName }) => {
+    const existingAccountName = accounts.some(
+      (acc) => acc.metadata.name.toLowerCase() === accountName.toLowerCase()
+    );
+    if (existingAccountName) {
+      setError('accountName', { message: `Account with name "${accountName}" already exists` });
```
Comment (ManuGowda) [inline src/modules/account/components/EditAccount/EditAccountForm.js]:
```suggestion
      setError('accountName', { message: `Account with name "${accountName}" already exists.` });
```

No translation?

File: src/modules/bookmark/utils.js [hunk 6]
```diff
@@ -26,17 +25,25 @@ export const validateBookmarks = (data) => {
 export const getIndexOfBookmark = (bookmarks, { address, token = tokenMap.LSK.key }) =>
   bookmarks[token].findIndex((bookmark) => bookmark.address === address);
 
+export const getIndexOfLabel = (bookmarks, { label, token = tokenMap.LSK.key }) =>
+  bookmarks[token].findIndex((item) => item.title === label);
+
```
Comment (ManuGowda) [inline src/modules/bookmark/utils.js]: The usage of `token` is wrong? `tokenMap.LSK.key` this is hardcoding, also we never need to store bookmark specific to any token? its a old code..

File: src/modules/bookmark/utils.js (hunk 3 above)
Comment (ManuGowda) [reply src/modules/bookmark/utils.js:41]: Refer my above comment.. we shouldn't store bookmark by token.. they are generic?

File: src/modules/bookmark/utils.js [hunk 7]
```diff
@@ -26,17 +25,25 @@ export const validateBookmarks = (data) => {
 export const getIndexOfBookmark = (bookmarks, { address, token = tokenMap.LSK.key }) =>
   bookmarks[token].findIndex((bookmark) => bookmark.address === address);
 
+export const getIndexOfLabel = (bookmarks, { label, token = tokenMap.LSK.key }) =>
```
Comment (ManuGowda) [inline src/modules/bookmark/utils.js:29]: why export a function when its not re used anywhere?

**Review by ManuGowda — CHANGES_REQUESTED** [review]
Move the position

Screenshot 2023-07-13 at 10 51 17 PM

Apply same validation to bookmark, user should not be allowed to enter space.

Screenshot 2023-07-13 at 10 57 02 PM
Screenshot 2023-07-13 at 10 57 25 PM

File: src/modules/bookmark/utils.js (hunk 7 above)
Comment (ikem-legend) [reply src/modules/bookmark/utils.js:29]: The export is used in the test

File: src/modules/account/components/EditAccount/EditAccountForm.js (hunk 5 above)
Comment (ikem-legend) [reply src/modules/account/components/EditAccount/EditAccountForm.js]: Fixed

**Review by ManuGowda — COMMENTED** [review]
Screenshot 2023-07-14 at 1 37 02 PM

Fixed

File: src/modules/auth/components/SetPasswordForm/SetPasswordForm.js [hunk 8]
```diff
@@ -64,6 +66,15 @@ function SetPasswordForm({ prevStep, onSubmit, recoveryPhrase, customDerivationP
   );
 
   const onFormSubmit = async (values) => {
+    const existingAccountName = accounts.some(
+      (acc) => acc.metadata.name.toLowerCase() === values.accountName.toLowerCase()
+    );
+    if (values.accountName && existingAccountName) {
+      setError('accountName', {
+        message: `Account with name "${values.accountName}" already exists`,
```
Comment (oskarleonard) [inline src/modules/auth/components/SetPasswordForm/SetPasswordForm.js]: translation?

Comment (ikem-legend) [reply src/modules/auth/components/SetPasswordForm/SetPasswordForm.js]: Fixed

File: setup/jest.config.js (hunk 1 above)
Comment (oskarleonard) [inline setup/jest.config.js:132]: Edit bookmark allows me to remove the name and to remove it for many bookmarks as well
image

Comment (oskarleonard) [inline setup/jest.config.js:132]: I can still get duplicate network by clicking edit and then save
image

Comment (ikem-legend) [reply setup/jest.config.js:132]: Fixed

File: src/modules/bookmark/utils.js [hunk 9]
```diff
@@ -26,17 +26,32 @@ export const validateBookmarks = (data) => {
...
+    return t('Label can be alphanumeric with either !,@,$,&,_,. as special characters');
+  }
+  if (value.length < 3) {
+    return t('Label is too short, Min. 3 characters');
+  }
   if (value.length > 20) {
     return t('Label is too long, Max. 20 characters');
   }
+  if (getIndexOfLabel(bookmarks, { label: value, token }) !== -1) {
+    return t(`Bookmark with name "${value}" already exists`);
```
Comment (ManuGowda) [inline src/modules/bookmark/utils.js]:
```suggestion
    return t(`Bookmark with name "${value}" already exists.`);
```

File: src/modules/bookmark/utils.test.js [hunk 10]
```diff
@@ -36,5 +53,93 @@ describe('Bookmarks', () => {
...
+      const value = 'really_long_bookmark_name';
+      expect(validateBookmarkLabel('LSK', value, accounts, t)).to.equal(
+        'Label is too long, Max. 20 characters'
+      );
+    });
+
+    it('returns feedback for duplicate labels', () => {
+      const value = 'lisker';
+      expect(validateBookmarkLabel('LSK', value, accounts, t)).to.equal(
+        `Bookmark with name "${value}" already exists`
```
Comment (ManuGowda) [inline src/modules/bookmark/utils.test.js]:
```suggestion
        `Bookmark with name "${value}" already exists.`
```

File: src/modules/bookmark/components/AddBookmark/AddBookmark.test.js [hunk 11]
```diff
@@ -136,14 +151,41 @@ describe('Add a new bookmark component', () => {
...
+          .simulate('change', {
+            target: {
+              value: accounts.validator.summary.address,
+              name: 'address',
+            },
+          });
+        expect(wrapper.find('input[name="label"]')).toHaveClassName('error');
+        expect(wrapper).toContainMatchingElement('.error');
+        expect(wrapper.find('.feedback.error')).toHaveText(
+          `Bookmark with name "${bookmarkTitle}" already exists`
```
Comment (ManuGowda) [inline src/modules/bookmark/components/AddBookmark/AddBookmark.test.js]:
```suggestion
          `Bookmark with name "${bookmarkTitle}" already exists.`
```

**Review by ManuGowda — DISMISSED** [review]

File: src/modules/network/components/DialogAddNetwork/DialogAddNetwork.js [hunk 12]
```diff
@@ -54,13 +54,27 @@ const DialogAddNetwork = () => {
...
+      !isNetworkServiceUrlChanged;
+    const isValidInEditMode =
+      isValidNetworkNameInEditMode && isValidNetworkServiceUrlChangeInEditMode;
+
+    if (!defaultName && !isValidInAddMode) {
       setErrorText('Network name or serviceUrl already exists');
       return;
     }
+    if (!!defaultName && !isValidInEditMode) {
+      setErrorText('Network name or serviceUrl already exists');
```
Comment (ManuGowda) [inline src/modules/network/components/DialogAddNetwork/DialogAddNetwork.js]: add the missing period.

Comment (ikem-legend) [reply src/modules/network/components/DialogAddNetwork/DialogAddNetwork.js]: Fixed

**Review by ManuGowda — DISMISSED** [review]

**Review by ManuGowda — APPROVED** [review]

**Review by oskarleonard — APPROVED** [review]

File: setup/jest.config.js (hunk 1 above)
Comment (ikem-legend) [reply setup/jest.config.js:132]: Done

File: setup/jest.config.js (hunk 2 above)
Comment (ikem-legend) [reply setup/jest.config.js:133]: Fixed

**Comment by ManuGowda** [comment]
The remaining improvements will be made in https://github.com/LiskHQ/lisk-desktop/issues/5184

File: src/modules/bookmark/utils.js (hunk 3 above)
Comment (ikem-legend) [reply src/modules/bookmark/utils.js:41]: To be done in a different PR

File: src/modules/bookmark/utils.js (hunk 6 above)
Comment (ikem-legend) [reply src/modules/bookmark/utils.js]: To be done in a separate PR

**Pull request merged**